examples/shamir-secret-share.masm

```diff
@@ -0,0 +1,150 @@
+#! This is a demo and is not secure.
+#!
+#! Input:
+#! Stack: [x, k] 
+#! The x coordinate for which we want to generate a share, number of shares  needed to 
+#! reconstruct secret - this defines the degree of the polynomial.
+#! Advice Tape: [secret, rand]
+#! The secret, a random field element.
+#!
+#! Output:
+#! Stack: [x_coord, y_coord, com_1, com_2, com_3, com_4]
+#! The share x coordinate, the share y coordinate, secret commitment 
+#!
+#! Suggested Improvements:
+#! - Currently only takes a single field element as a random seed - should be increased to a word.
+#! - Migrate to a more secure random number generation algorithm for polynomial coefficient generation.
+#! - Allow arbitrary secret size of size n.  This would require generation of n polynomials.
+#! - Instead of computing x^i for each iteration, store x^i in a variable and multiply by x to get next exponent.
+
+
+# create secret commitment (hash of secret + rand)
+proc.gen_secret_com
+    # input: [secret, rand]
+    padw          #[0, 0, 0, 0, secret, rand]
+    padw          #[0, 0, 0, 0, 0, 0, 0, 0, secret, rand]
+    drop          #[0, 0, 0, 0, 0, 0, 0, secret, rand]
+    drop          #[0, 0, 0, 0, 0, 0, secret, rand]
+    rphash        #[com_1, com_2, com_3, com_4]
+end
+
+# create k-1 random polynomial coefficients
+proc.gen_poly_coeffs
+    # input [k, rand]
+
+    # generate initial 4 random coefficients from rand
+    swap    #[rand, k]
+    padw    #[0, 0, 0, 0, rand, k]
+    padw    #[0, 0, 0, 0, 0, 0, 0, 0, rand, k]
+    drop    #[0, 0, 0, 0, 0, 0, 0, rand, k]
+    rphash  #[coef_1, coef_2, coef_3, coef_4, k]
+
+    # introduce counter and evaluate if we have enough cofficients
+    push.5  #[counter, coef_1, coef_2, coef_3, coef_4, k]
+    movup.5 #[k, counter, coef_1, coef_2, coef_3, coef_4]
+    dup.1
+    dup.1
+    lt
+
+    # continue creating more coefficients until we have enough (created in 4's)
+    while.true     #[k, counter, diget_1, diget_2, digest_3, digest_4]
+        # generate more cofficients by padding and hashing previous cofficients
+        movdn.5
+        movdn.5
+        dupw       #[digest_1, digest_2, digest_3, digest_4, k, counter, ...]
+        padw       #[0, 0, 0, 0, diget_1, diget_2, digest_3, digest_4, diget_1, diget_2, digest_3, digest_4, k, counter]
+        rphash     #[digest_5, digest_6, digest_7, digest_8, digest_1, digest_2, digest_3, digest_4, k, counter, ...]
+
+        # evaluate if we have enugh coeficients
+        movup.9
+        add.4
+        movup.9
+        dup.1
+        dup.1
+        lt
+    end         #[k, counter, coef_1, ...]  
+    
+    # evaluate if we have too many cofficients
+    dup          #[k, k, counter, coef_1, ...]
+    movdn.2      #[k, counter, k, coef_1, ...]
+    sub          #[diff, k, coef_1, ...]
+    dup          #[diff, diff, k, coef_1, ...]
+    push.0       #[0, diff, diff, k, coef_1, ...]
+    gt           #[1, diff, k, coef_1] 
```

Comment:
mega nit: I don't think it is necessary to rewrite this, but it may be cool to share this strategy. Since the loop above produces `4` coefficients per iteration, it would be nice if `k` was a multiple of it. That can be done with something like:

```
               # stack: [e, ...]
u32cast        # stack: [low, ...] discard high bits
push.3         # stack: [4, low, ...] bit pattern for the lower bits
u32checked_and # stack: [diff, ...] the remainder
```

It is just `3` cycles to get the remainder. With it you can compute the number of loops:

```
               # stack: [k, ...]
dup.0          # stack: [k, k, ...]
exec.remainder # stack: [r, k, ...]
dup.1 dup.1    # stack: [r, k, r, k, ...]
sub            # stack: [loop, r, k, ...]
```

Later you set counter to `0` and increment it by `4` until it is equal to `loop`. This can be done with a `neq` check which is one cycle, instead of the `lt` which is `17`.

The remainder can be handle on a similar fashion, first get the number of elements to discard `4-r`, then set a counter variable to `0` and in the loop 1. drop the extra coef 2. increment the counter 3. do a `neq` check. This removes the `gt` which is `18` cycles in favor of a `neq` which is only one!

I think it is a neat trick to optimize cycle count :)

Reply:
Awesome trick, thank you for sharing!  When I get some free time I will implement this as it'll be a good way to practice my assembly optimisation.